lib/kube/proxy/forwarder.go

```diff
@@ -2148,3 +2312,358 @@ func (f *Forwarder) removeKubeDetails(name string) {
+	if err := filterBuffer(
+		newPodFiltererBuilder(allowedResources, deniedResources, f.log),
+		memBuffer,
+	); err != nil {
+		return memBuffer.Status(), trace.Wrap(err)
+	}
+	// Copy the filtered payload into target http.ResponseWriter.
+	err := memBuffer.CopyInto(w)
+
+	// Returns the status and any filter error.
+	return memBuffer.Status(), trace.Wrap(err)
+}
+
+// listPodsWatcher handles a long lived connection to the upstream server where
+// the Kubernetes API returns frames with events.
+// This handler creates a WatcherResponseWriter that spins a new goroutine once
+// the API server writes the status code and headers.
+// The goroutine waits for new events written into the response body and
+// decodes each event. Once decoded, we validate if the Pod name matches
+// any Pod specified in `kubernetes_resources` and if included, the event is
+// forwarded to the user's response writer.
+// If it does not match, the watcher ignores the event and continues waiting
+// for the next event.
+func (f *Forwarder) listPodsWatcher(req *http.Request, w http.ResponseWriter, sess *clusterSession, allowedResources, deniedResources []types.KubernetesResource) (int, error) {
+	negotiator := newClientNegotiator()
+	rw, err := responsewriters.NewWatcherResponseWriter(w, negotiator, newPodFiltererBuilder(allowedResources, deniedResources, f.log))
+	if err != nil {
+		return http.StatusInternalServerError, trace.Wrap(err)
+	}
+	// Forwards the request to the target cluster.
+	sess.forwarder.ServeHTTP(rw, req)
+	// Once the request terminates, close the watcher and waits for resources
+	// cleanup.
+	err = rw.Close()
+	return rw.Status(), trace.Wrap(err)
+}
+
+// deletePodsCollection calls checks the listPodsList
+func (f *Forwarder) deletePodsCollection(ctx *authContext, w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
+	sess, err := f.newClusterSession(*ctx)
+	if err != nil {
+		// This error goes to kubernetes client and is not visible in the logs
+		// of the teleport server if not logged here.
+		f.log.Errorf("Failed to create cluster session: %v.", err)
+		return nil, trace.Wrap(err)
+	}
+
+	sess.upgradeToHTTP2 = true
+	sess.forwarder, err = f.makeSessionForwarder(sess)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := f.setupForwardingHeaders(sess, req); err != nil {
+		// This error goes to kubernetes client and is not visible in the logs
+		// of the teleport server if not logged here.
+		f.log.Errorf("Failed to set up forwarding headers: %v.", err)
+		return nil, trace.Wrap(err)
+	}
+	// status holds the returned response code.
+	var status int
+	// Check if the target Kubernetes cluster is not served by the current service.
+	// If it's the case, forward the request to the target Kube Service where the
+	// filtering logic will be applied.
+	if !f.isLocalKubeCluster(sess) {
+		rw := httplib.NewResponseStatusRecorder(w)
+		sess.forwarder.ServeHTTP(rw, req)
+		status = rw.Status()
```

Comment:
There seems to be a common boilerplate between listPods and deletePodsCollection, and it appears to be almost identical up to this point. Should we refactor so we pull out the common parts?